[LineEdit] Implement input mask support.

bruvzg opened the pull request:
Adds support for input mask to the `LineEdit`, similar to the [Qt feature](https://doc.qt.io/qt-6/qlineedit.html#inputMask-prop) with the same name (and using the same mask format).

https://github.com/user-attachments/assets/8b10469e-a899-4091-b787-6a5a9118e119

Closes https://github.com/godotengine/godot-proposals/issues/7193

bruvzg committed:
[LineEdit] Implement input mask support.

Mickeon commented on doc/classes/LineEdit.xml:
```diff
@@ -283,6 +283,23 @@
 		</member>
 		<member name="focus_mode" type="int" setter="set_focus_mode" getter="get_focus_mode" overrides="Control" enum="Control.FocusMode" default="2" />
+		<member name="input_mask" type="String" setter="set_input_mask" getter="get_input_mask" default="&quot;&quot;">
+			The text input mask. Input mask is a template string composed of the following characters:
+			- [code]A[/code] - mask for ASCII alphabet character, [code]A-Z, a-z[/code].
+			- [code]N[/code] - mask for ASCII alphanumeric character, [code]0-9, A-Z, a-z[/code].
+			- [code]9[/code] - mask for ASCII number character, [code]0-9[/code].
+			- [code]D[/code] - mask for ASCII number character larger than zero, [code]1-9[/code].
+			- [code]#[/code] - mask for plus/minus sign, [code]+, -[/code].
+			- [code]H[/code] - mask for hexadecimal number character, [code]0-9, A-F, a-f[/code].
+			- [code]B[/code] - mask for binary number character, [code]0, 1[/code].
+			- [code]X[/code] - mask for any other character.
+			- [code]&gt;[/code] - control character, all following characters are converted to upper case.
+			- [code]&lt;[/code] - control character, all following characters are converted to lower case.
+			- [code]![/code] - control character, disable case conversion for all following characters.
+			- [code];[/code] - control character, followed by the single character, terminates the mask and sets blank character (default blank character is space).
+			- [code]\[/code] - control character, used to escape mask and control characters.
+			- All other characters are used as separators.
```
```suggestion
			The text input mask. This is a template string composed of the following characters:
			- [code]A[/code] - mask for an ASCII alphabet character ([code]A-Z[/code], [code]a-z[/code]).
			- [code]N[/code] - mask for an ASCII alphanumeric character ([code]0-9[/code], [code]A-Z[/code], [code]a-z[/code]).
			- [code]9[/code] - mask for an ASCII number character ([code]0-9[/code]).
			- [code]D[/code] - mask for an ASCII number character greater than zero ([code]1-9[/code]).
			- [code]#[/code] - mask for a plus/minus sign ([code]+[/code], [code]-[/code]).
			- [code]H[/code] - mask for a hexadecimal number character ([code]0-9[/code], [code]A-F[/code], [code]a-f[/code]).
			- [code]B[/code] - mask for a binary number character ([code]0[/code], [code]1[/code]).
			- [code]X[/code] - mask for any other character.
			- [code]&gt;[/code] - control character, all following characters are converted to uppercase.
			- [code]&lt;[/code] - control character, all following characters are converted to lowercase.
			- [code]![/code] - control character, disable case conversion for all following characters.
			- [code];[/code] - control character, when followed by the single character, terminates the mask and sets the blank character  (the default blank character is space).
			- [code]\[/code] - control character, used to escape the mask and control characters.
			- All other characters are used as separators.
```

Mickeon commented on doc/classes/LineEdit.xml:
```diff
@@ -316,6 +333,9 @@
 		</member>
 		<member name="mouse_default_cursor_shape" type="int" setter="set_default_cursor_shape" getter="get_default_cursor_shape" overrides="Control" enum="Control.CursorShape" default="1" />
+		<member name="overtype_mode_enabled" type="bool" setter="set_overtype_mode_enabled" getter="is_overtype_mode_enabled" default="false">
+			If [code]true[/code], sets the user into overtype mode. When the user types in this mode, it will override existing text.
```
Copied straight from a revamped **TextEdit** description.
```suggestion
			If [code]true[/code], enables overtype mode. In this mode, typing overrides existing text instead of inserting text. The [member ProjectSettings.input/ui_text_toggle_insert_mode] action toggles overtype mode.
```

Mickeon commented on doc/classes/LineEdit.xml:
```diff
@@ -517,6 +537,9 @@
 			Default font color.
 		</theme_item>
+		<theme_item name="font_mask_color" data_type="color" type="Color" default="Color(0.875, 0.875, 0.875, 0.6)">
+			Font color for static [member input_mask] parts.
```
Maybe...?
```suggestion
			Font color for the static characters composing the [member input_mask].
```

Mickeon reviewed (commented):
Nice feature, but I don't see this being exposed due to general lack of interest.
It may be good to have _somewhere_ in the editor itself, perhaps as a PROPERTY_HINT constant?

Calinou reviewed (commented):
Tested locally, it mostly works as expected.

**Testing project:** [test_pr_98727.zip](https://github.com/user-attachments/files/18011357/test_pr_98727.zip)

I noticed some issues:

- You can freely place your cursor around before writing in, which means you can accidentally start writing after the first character (e.g. with the mask `>AAAAA-AAAAA-AAAAA-AAAAA`):

![image](https://github.com/user-attachments/assets/a3c16c3f-ed86-4db6-b1f3-ca256d7fdac8)

This may be convenient in certain cases, but it's also prone to misinputs, so I'd suggest not allowing placing the cursor within whitespace.

- With the input mask `+99 99 99 99 99 99`, I can't enter the last character:

![image](https://github.com/user-attachments/assets/d3b3b013-520e-4a98-9759-293e539514ee)

This also occurs with the `>AAAAA-AAAAA-AAAAA-AAAAA` input mask, where the last section is limited to 4 characters instead of 5.

It works with the mask `+99 99 99 99 99 99;_` (or even `+99 99 99 99 99 99; ` with a trailing space), as setting a custom blank character fixes the issue (`+99 99 99 99 99 99;` doesn't work either). I can now enter the final digit:

![image](https://github.com/user-attachments/assets/fadc30da-f2a1-4e71-bfff-4c593abd6092)

bruvzg commented:
> input mask, where the last section is limited to 4 characters instead of 5.
> It works with the mask +99 99 99 99 99 99;_ (or even +99 99 99 99 99 99; with a trailing space), as setting a custom blank character fixes the issue (+99 99 99 99 99 99; doesn't work either). 

Fixed.

> You can freely place your cursor around before writing in, which means you can accidentally start writing after the first character

Not sure about this, I have changed it to always reset caret on focus, but editing specific part might be useful (also nothing in the mask indicate that entering all character is required).

Calinou reviewed (approved):
Works great now 🙂

AThousandShips commented on scene/gui/line_edit.cpp:
```diff
@@ -1652,6 +1717,165 @@ Vector2 LineEdit::get_caret_pixel_pos() {
+
+String LineEdit::_apply_mask(const String &p_text) const {
+	String new_text;
+	new_text.resize(mask.size() + 1);
```
```suggestion
	new_text.resize(mask.size() + 1);
	char32_t *new_text_ptr = new_text.ptrw();
```

AThousandShips commented on scene/gui/line_edit.cpp:
```diff
@@ -1652,6 +1717,165 @@ Vector2 LineEdit::get_caret_pixel_pos() {
+		} else if (e.char_case == MASK_CASE_LOWER) {
+			c = String::char_lowercase(c);
+		}
+		new_text[i] = c;
```
```suggestion
		new_text_ptr[i] = c;
```

AThousandShips reviewed (commented):
Code looks good, will test when I can

akien-mga commented:
Needs rebase.

Mickeon commented:
I know it has been approved, but I would still like to reiterate that it's quite the niche to have on the base **LineEdit**. Let's be a bit careful on adding new features.
